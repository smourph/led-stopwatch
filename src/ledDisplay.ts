const LED_PATTERNS: number[][] = [
  [1, 1, 1, 1, 1, 1, 0], // 0
  [0, 1, 1, 0, 0, 0, 0], // 1
  [1, 1, 0, 1, 1, 0, 1], // 2
  [1, 1, 1, 1, 0, 0, 1], // 3
  [0, 1, 1, 0, 0, 1, 1], // 4
  [1, 0, 1, 1, 0, 1, 1], // 5
  [1, 0, 1, 1, 1, 1, 1], // 6
  [1, 1, 1, 0, 0, 0, 0], // 7
  [1, 1, 1, 1, 1, 1, 1], // 8
  [1, 1, 1, 1, 0, 1, 1], // 9
] as const;

type Variant = 'standard' | 'small' | 'bottom' | 'small-bottom';

export function renderLedDigit(digit: string, variant: Variant = 'standard'): string {
  if (digit === ':') {
    return '<div class="led-colon">'
      + '<div></div>'
      + '<div></div>'
      + '</div>';
  }

  if (digit === '.') {
    return '<div class="led-dot led-dot--small led-dot--bottom"></div>';
  }

  return renderNumberDigit();

  function renderNumberDigit(): string {
    const n: number = parseInt(digit, 10);
    if (isNaN(n)) return '';

    const pattern = LED_PATTERNS[n];

    const classes = [
      'led-digit',
      ...[
        variant === 'small' || variant === 'small-bottom' ? 'led-digit--small' : undefined,
        variant === 'bottom' || variant === 'small-bottom' ? 'led-digit--bottom' : undefined,
      ].filter(Boolean) as string[]
    ].join(' ');

    return `<div class="${classes}">`
      + `<div class="segment horizontal top${pattern[0] ? ' on' : ''}"></div>`
      + `<div class="segment vertical top-right${pattern[1] ? ' on' : ''}"></div>`
      + `<div class="segment vertical bottom-right${pattern[2] ? ' on' : ''}"></div>`
      + `<div class="segment horizontal bottom${pattern[3] ? ' on' : ''}"></div>`
      + `<div class="segment vertical bottom-left${pattern[4] ? ' on' : ''}"></div>`
      + `<div class="segment vertical top-left${pattern[5] ? ' on' : ''}"></div>`
      + `<div class="segment horizontal middle${pattern[6] ? ' on' : ''}"></div>`
      + `</div>`;
  }
}

export function renderLedTime(elapsed: number): string {
  const timeStr: string = formatTimeString(elapsed);
  return Array.from(timeStr)
    .map((char, idx) =>
      idx >= 6 // Last 3 chars are milliseconds â†’ use small-bottom
        ? renderLedDigit(char, 'small-bottom')
        : renderLedDigit(char))
    .join('');
}

export function formatTimeString(elapsed: number): string {
  const maxElapsed: number = 59 * 60 * 1000 + 59 * 1000 + 999;
  const safeElapsed: number = Math.min(elapsed, maxElapsed);
  const ms: number = Math.floor(safeElapsed % 1000);
  const s: number = Math.floor((safeElapsed / 1000) % 60);
  const m: number = Math.floor((safeElapsed / 60000) % 60);
  return `${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}.${ms.toString().padStart(3, '0')}`;
}
