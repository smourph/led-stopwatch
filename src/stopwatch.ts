export class Stopwatch {
  private startTime: number | null = null;
  private elapsedBeforePause: number = 0;
  private animationFrameId: number | null = null;
  private readonly renderCallback: (elapsed: number) => void;
  private readonly boundUpdate: () => void;
  private static readonly MAX_ELAPSED: number = 59 * 60 * 1000 + 59 * 1000 + 999;

  constructor(renderCallback: (elapsed: number) => void) {
    this.renderCallback = renderCallback;
    this.boundUpdate = this.update.bind(this); // bind once
  }

  public start(): void {
    if (this.isRunning()) return;

    this.startTime = performance.now();
    this.animationFrameId = requestAnimationFrame(this.boundUpdate);
  }

  public pause(): void {
    if (!this.isRunning()) return;

    this.elapsedBeforePause += performance.now() - (this.startTime as number);
    this.startTime = null;
    this.cancelAnimationFrame();
  }

  public reset(): void {
    this.startTime = null;
    this.elapsedBeforePause = 0;
    this.cancelAnimationFrame();
    this.renderCallback(0);
  }

  private update(): void {
    if (!this.isRunning()) return;

    const now: number = performance.now();
    let elapsed: number = this.elapsedBeforePause + (now - (this.startTime as number));
    if (elapsed > Stopwatch.MAX_ELAPSED) {
      elapsed = Stopwatch.MAX_ELAPSED;
      this.pause();
    }

    this.renderCallback(elapsed);
    this.animationFrameId = requestAnimationFrame(this.boundUpdate);
  }

  private cancelAnimationFrame(): void {
    if (this.animationFrameId !== null) {
      cancelAnimationFrame(this.animationFrameId);
      this.animationFrameId = null;
    }
  }

  public isRunning(): boolean {
    return this.startTime !== null;
  }

  public destroy(): void {
    this.cancelAnimationFrame();
    this.startTime = null;
    this.elapsedBeforePause = 0;
  }
}