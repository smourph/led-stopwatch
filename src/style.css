/* Root variables for easy tuning */
:root {
  --led-width: 7.5em;
  --led-height: 15em;
  --led-gap: 1.2em;
  --dot-size: 1.2em;
  --small-scale: 0.40;
}

body {
  display: flex;
  align-items: center;
  justify-content: center;
  font-family: monospace;
  background: #f7f7fa;
  height: 100vh;
  margin: 0;
  padding: 0;
}

#stopwatch-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
}

#stopwatch {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 1em 2em;
  margin-bottom: 3em;
}

.btn {
  padding: 1rem 3rem;
  line-height: 1.5;
  border: 1px solid transparent;
  border-radius: 0.75em;
  font-size: 3.5rem;
  font-weight: bold;
  transition: color .15s ease-in-out, background-color .15s ease-in-out, border-color .15s ease-in-out, box-shadow .15s ease-in-out;
  user-select: none;
  cursor: pointer;
}

.reset-btn {
  color: #fff;
  background-color: #007bff;
  border-color: #007bff;
}

.reset-btn:hover, .reset-btn:focus {
  text-decoration: none;
  background-color: #0069d9;
  border-color: #0062cc;
}

.led-digit {
  position: relative;
  width: var(--led-width);
  height: var(--led-height);
  display: inline-block;
  margin: 0 var(--led-gap);
}

.led-digit--small {
  width: calc(var(--led-width) * var(--small-scale));
  height: calc(var(--led-height) * var(--small-scale));
  margin: 0 calc(var(--led-gap) * var(--small-scale));
}

.led-digit--bottom {
  align-self: flex-end;
  margin-top: auto;
  margin-bottom: 0.2em;
}

.segment {
  position: absolute;
  background: transparent;
  border-radius: 0.7em;
  transition: background 0.2s;
}

.segment.vertical {
  width: 14%;
  height: 42%;
}

.segment.horizontal {
  width: 76%;
  height: 7%;
  left: 12%;
}

.segment.top {
  top: 0;
}

.segment.bottom {
  bottom: 0;
}

.segment.middle {
  top: 46.5%;
}

.segment.top-right, .segment.bottom-right {
  right: 0;
}

.segment.bottom-left, .segment.top-left {
  left: 0;
}

.segment.top-right, .segment.top-left {
  top: 6%;
}

.segment.bottom-right, .segment.bottom-left {
  bottom: 6%;
}

.segment.on {
  background: #111217;
  box-shadow: 0 0 16px rgba(17, 18, 23, 0.6);
}

.led-colon, .led-dot {
  width: 2.2em;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.led-colon div, .led-dot {
  width: calc(var(--dot-size));
  height: calc(var(--dot-size));
  border-radius: 50%;
  background: #111217;
  box-shadow: 0 0 16px rgba(17, 18, 23, 0.6);
}

.led-colon div {
  margin: 0.5em 0;
}

.led-dot {
  margin-top: 2.5em;
}

.led-dot--small {
  width: calc(var(--dot-size) * 0.666);
  height: calc(var(--dot-size) * 0.666);
}

.led-dot--bottom {
  align-self: flex-end;
  margin-top: auto;
  margin-bottom: 0.2em;
}
