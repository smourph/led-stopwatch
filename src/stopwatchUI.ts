import {Stopwatch} from './stopwatch';
import {formatTimeString, renderLedTime} from './ledDisplay';

export class StopwatchUI {
  private readonly stopwatch: Stopwatch;
  private readonly stopwatchDiv: HTMLElement;
  private readonly container: HTMLElement;
  private readonly resetBtn: HTMLButtonElement | null;

  private readonly boundOnContainerClick: (e: MouseEvent) => void;
  private readonly boundOnResetClick: (e: MouseEvent) => void;
  private readonly boundOnKeyDown: (e: KeyboardEvent) => void;

  constructor() {
    const container = document.getElementById('stopwatch-container');
    const stopwatchEl = document.getElementById('stopwatch');

    if (!container || !stopwatchEl) {
      throw new Error('StopwatchUI: required DOM elements not found (#stopwatch-container, #stopwatch)');
    }

    this.container = container;
    this.stopwatchDiv = stopwatchEl;

    const reset = document.getElementById('reset');
    this.resetBtn = reset instanceof HTMLButtonElement ? reset : null;

    this.stopwatch = new Stopwatch(this.render.bind(this));

    this.boundOnContainerClick = this.onContainerClick.bind(this);
    this.boundOnResetClick = this.onResetClick.bind(this);
    this.boundOnKeyDown = this.onKeyDown.bind(this);

    this.attachEvents();
    this.render(0);
  }

  private render(elapsed: number): void {
    const timeStr: string = formatTimeString(elapsed);

    this.stopwatchDiv.innerHTML = renderLedTime(elapsed);

    this.stopwatchDiv.setAttribute('data-time', timeStr);
    this.stopwatchDiv.setAttribute('aria-label', timeStr);
  }

  private onContainerClick(): void {
    if (this.stopwatch.isRunning()) {
      this.stopwatch.pause();
    } else {
      this.stopwatch.start();
    }
  }

  private onResetClick(e: MouseEvent): void {
    e.stopPropagation();
    this.stopwatch.reset();
  }

  private onKeyDown(e: KeyboardEvent): void {
    // Space toggles start/pause, Escape resets
    if (e.code === 'Space' || e.key === ' ') {
      e.preventDefault();
      this.stopwatch.isRunning() ? this.stopwatch.pause() : this.stopwatch.start();
    } else if (e.key === 'Escape') {
      this.stopwatch.reset();
    }
  }

  private attachEvents(): void {
    this.container.addEventListener('click', this.boundOnContainerClick);
    if (this.resetBtn) this.resetBtn.addEventListener('click', this.boundOnResetClick);
    document.addEventListener('keydown', this.boundOnKeyDown);
  }

  public destroy(): void {
    this.container.removeEventListener('click', this.boundOnContainerClick);
    if (this.resetBtn) this.resetBtn.removeEventListener('click', this.boundOnResetClick);
    document.removeEventListener('keydown', this.boundOnKeyDown);
    this.stopwatch.destroy();
  }
}
