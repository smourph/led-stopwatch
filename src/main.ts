import {StopwatchUI} from './stopwatchUI';

export function bootstrap(): StopwatchUI {
  return new StopwatchUI();
}

// Auto-start when loaded in browser and keep reference to instance for lifecycle
if (typeof document !== 'undefined') {
  let _instance: StopwatchUI | null = null;
  const start = () => {
    _instance = bootstrap();
    // destroy on unload to clean up event listeners
    window.addEventListener('beforeunload', () => {
      _instance?.destroy();
      _instance = null;
    });
  };

  if (document.readyState === 'loading') {
    window.addEventListener('DOMContentLoaded', start);
  } else {
    start();
  }
}
