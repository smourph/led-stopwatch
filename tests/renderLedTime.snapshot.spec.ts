import {describe, expect, it} from 'vitest';
import {renderLedTime} from '../src/ledDisplay';

describe('renderLedTime snapshot', () => {
  it('renders full time (12:34.567) consistently', () => {
    const html = renderLedTime((12 * 60 + 34) * 1000 + 567);
    expect(html).toMatchSnapshot();
  });

  it('renders zero time (00:00.000) consistently', () => {
    const html = renderLedTime(0);
    expect(html).toMatchSnapshot();
  });
});

