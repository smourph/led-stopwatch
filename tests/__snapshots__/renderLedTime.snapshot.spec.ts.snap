// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderLedTime snapshot > renders full time (12:34.567) consistently 1`] = `"<div class="led-digit"><div class="segment horizontal top"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom"></div><div class="segment vertical bottom-left"></div><div class="segment vertical top-left"></div><div class="segment horizontal middle"></div></div><div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left"></div><div class="segment horizontal middle on"></div></div><div class="led-colon"><div></div><div></div></div><div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left"></div><div class="segment vertical top-left"></div><div class="segment horizontal middle on"></div></div><div class="led-digit"><div class="segment horizontal top"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom"></div><div class="segment vertical bottom-left"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle on"></div></div><div class="led-dot led-dot--small led-dot--bottom"></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle on"></div></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle on"></div></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom"></div><div class="segment vertical bottom-left"></div><div class="segment vertical top-left"></div><div class="segment horizontal middle"></div></div>"`;

exports[`renderLedTime snapshot > renders zero time (00:00.000) consistently 1`] = `"<div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-colon"><div></div><div></div></div><div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-digit"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-dot led-dot--small led-dot--bottom"></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div><div class="led-digit led-digit--small led-digit--bottom"><div class="segment horizontal top on"></div><div class="segment vertical top-right on"></div><div class="segment vertical bottom-right on"></div><div class="segment horizontal bottom on"></div><div class="segment vertical bottom-left on"></div><div class="segment vertical top-left on"></div><div class="segment horizontal middle"></div></div>"`;
